### PR TITLE
Db upgrade

### PR DESCRIPTION
**Pull request by aaronbuchwald**
This PR adds ChainState to provide an efficient caching layer for blocks that sits on top of the VM. This PR requires a database upgrade.

### CODE REVIEW

File: plugin/evm/state/state.go [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+		}
+		if err := c.acceptedBlocksByHeight.Put(heightKey(genesisBlock.Height()), genesisID[:]); err != nil {
+			return err
+		}
+		lastAcceptedBlock = &BlockWrapper{
+			Block:  genesisBlock,
+			status: choices.Accepted,
+		}
+	}
+	lastAcceptedBlock.state = c
```
Comment (StephenButtolph) [inline plugin/evm/state/state.go:140]: Can we populate this field when we construct the `BlockWrapper`s?

File: plugin/evm/state/state.go [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+		return blk, nil
+	}
+
+	if _, ok := c.missingBlocks.Get(blkID); ok {
+		return nil, errUnknownBlock
+	}
+
+	c.lock.Unlock()
+	blk, err := c.getBlock(blkID)
+	c.lock.Lock()
```
Comment (StephenButtolph) [inline plugin/evm/state/state.go:176]: this looks really racy.

File: plugin/evm/state/state.go [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+		return nil, err
+	}
+
+	return wrappedBlk.(*BlockWrapper).Block, nil
+}
+
+// getBlockIDAtHeight returns the blkID accepted at [height] if there is
+// one.
+// assumes the lock is held.
+func (c *ChainState) getBlockIDAtHeight(height uint64) (ids.ID, error) {
```
Comment (StephenButtolph) [inline plugin/evm/state/state.go:218]: Why does this exist? Shouldn't this just get Geth's canonical hash?

File: plugin/evm/state/state_test.go [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+	"github.com/ava-labs/avalanchego/database/memdb"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/snow/choices"
+	"github.com/ava-labs/avalanchego/snow/consensus/snowman"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/stretchr/testify/assert"
+)
+
+type testBlock struct {
```
[the remark this answers is not in the document]
Comment (aaronbuchwald) [reply plugin/evm/state/state_test.go:18]: Yes, just wrapped it to provide a SetStatus function as well.

File: plugin/evm/state/state.go (hunk 3 above)
Comment (aaronbuchwald) [reply plugin/evm/state/state.go:218]: Going to remove in favor of geth's canonical hash.

File: plugin/evm/state/state.go (hunk 2 above)
Comment (aaronbuchwald) [reply plugin/evm/state/state.go:176]: Removing all of the locks.

**Review by StephenButtolph — APPROVED** [review]
I think it needs to have a `go mod tidy` run on it, but aside from that, LGTM

File: coreth.go [hunk 5]
```diff
@@ -62,9 +61,8 @@ func NewETHChain(config *eth.Config, nodecfg *node.Config, etherBase *common.Add
 	//}
 	cb := new(dummy.ConsensusCallbacks)
 	mcb := new(miner.MinerCallbacks)
-	bcb := new(eth.BackendCallbacks)
-	backend, _ := eth.New(node, config, cb, mcb, bcb, chainDB, settings)
-	chain := &ETHChain{backend: backend, cb: cb, mcb: mcb, bcb: bcb}
+	backend, _ := eth.New(node, config, cb, mcb, chainDB, settings)
```
Comment (patrick-ogrady) [inline coreth.go:64]: Why don't we check this error?

Comment (aaronbuchwald) [reply coreth.go:64]: Never has been before, but will update now to panic if there's an error here.

**Pull request closed without merging**